src/libstd/sys_common/backtrace.rs

```diff
@@ -25,45 +27,107 @@ pub const HEX_WIDTH: usize = 18;
+pub fn log_enabled() -> Option<PrintFormat> {
     static ENABLED: atomic::AtomicIsize = atomic::AtomicIsize::new(0);
     match ENABLED.load(Ordering::SeqCst) {
-        1 => return false,
-        2 => return true,
-        _ => {}
+        0 => {},
+        1 => return None,
+        2 => return Some(PrintFormat::Full),
+        3 => return Some(PrintFormat::Short),
+        _ => unreachable!(),
     }
 
     let val = match env::var_os("RUST_BACKTRACE") {
-        Some(x) => if &x == "0" { 1 } else { 2 },
-        None => 1,
+        Some(x) => if ["no", "0"].iter().any(|val| &x == *val) {
+            None
+        } else if &x == "full" {
+            Some(PrintFormat::Full)
+        } else if ["short", "yes", "1"].iter().any(|val| &x == *val)  {
+            Some(PrintFormat::Short)
+        } else {
+            // Backward compatibility
+            Some(PrintFormat::Short)
+        },
+        None => None,
     };
-    ENABLED.store(val, Ordering::SeqCst);
-    val == 2
+    ENABLED.store(match val {
+        Some(v) => v as isize,
+        None => 1,
+    }, Ordering::SeqCst);
+    val
 }
 
 // These output functions should now be used everywhere to ensure consistency.
 pub fn output(w: &mut Write, idx: isize, addr: *mut libc::c_void,
-              s: Option<&[u8]>) -> io::Result<()> {
-    write!(w, "  {:2}: {:2$?} - ", idx, addr, HEX_WIDTH)?;
-    match s.and_then(|s| str::from_utf8(s).ok()) {
-        Some(string) => demangle(w, string)?,
-        None => write!(w, "<unknown>")?,
+              s: Option<&str>, format: PrintFormat) -> io::Result<()> {
+    // Remove the `17: 0x0 - <unknown>` line.
+    if format == PrintFormat::Short && addr == ptr::null_mut() {
+        return Ok(());
+    }
+    write!(w, "  {:2}: ", idx)?;
+    match format {
+        PrintFormat::Full => write!(w, "{:1$?} - ", addr, HEX_WIDTH)?,
+        PrintFormat::Short => w.write_all(b"   ")?,
+    }
+    match s {
+        Some(string) => demangle(w, string, format)?,
+        None => w.write_all(b"<unknown>")?,
     }
-    w.write_all(&['\n' as u8])
+    w.write_all(b"\n")
 }
 
 #[allow(dead_code)]
 pub fn output_fileline(w: &mut Write, file: &[u8], line: libc::c_int,
-                       more: bool) -> io::Result<()> {
-    let file = str::from_utf8(file).unwrap_or("<unknown>");
+                       more: bool, format: PrintFormat) -> io::Result<()> {
     // prior line: "  ##: {:2$} - func"
-    write!(w, "      {:3$}at {}:{}", "", file, line, HEX_WIDTH)?;
+    w.write_all(b"           ")?;
+    match format {
+        PrintFormat::Full => write!(w, "{:1$}", "", HEX_WIDTH)?,
+        PrintFormat::Short => ()
+    }
+
+    let file = str::from_utf8(file).unwrap_or("<unknown>");
+    let file_path = Path::new(file);
+    let mut already_printed = false;
+    if format == PrintFormat::Short && file_path.is_absolute() {
+        if let Ok(cwd) = env::current_dir() {
+            let buildroot = Path::new(env!("CFG_BUILD_DIR"));
```

Comment:
Could this use `option_env` and handle the `None` case to ensure that we can build the standard library separately if need be?